specs/core/1_shard-data-chains.md

```diff
@@ -17,596 +17,400 @@
+### Block processing
 
-    assert bls_verify(
-        pubkey=bls_aggregate_pubkeys(pubkeys),
-        message_hash=block.core.parent_root,
-        signature=block.signatures.attestation_signature,
-        domain=get_domain(beacon_state, DOMAIN_SHARD_ATTESTER, compute_epoch_of_shard_slot(block.core.slot))
-    )
+```python
+def process_shard_block(beacon_state: BeaconState, shard_state: ShardState, block: ShardBlock) -> None:
+    process_shard_block_header(beacon_state, shard_state, block)
+    process_shard_attestations(beacon_state, shard_state, block)
+    process_shard_block_body(beacon_state, shard_state, block)
+```
 
-    # Check proposer
-    proposer_index = get_shard_block_proposer_index(beacon_state, state.shard, block.core.slot)
-    assert proposer_index is not None
-    add_reward(state, beacon_state, proposer_index, attestations * base_reward // PROPOSER_REWARD_QUOTIENT)
-    assert bls_verify(
-        pubkey=beacon_state.validators[proposer_index].pubkey,
-        message_hash=hash_tree_root(block.core),
-        signature=block.signatures.proposer_signature,
-        domain=get_domain(beacon_state, DOMAIN_SHARD_PROPOSER, compute_epoch_of_shard_slot(block.core.slot)),
-    )
+#### Block header
 
-    # Process and update block data fees
-    add_fee(state, beacon_state, proposer_index, state.basefee * len(block.core.data) // SHARD_BLOCK_SIZE_LIMIT)
-    QUOTIENT = SHARD_BLOCK_SIZE_LIMIT * BASEFEE_ADJUSTMENT_FACTOR
-    if len(block.core.data) > SHARD_BLOCK_SIZE_TARGET:
-        state.basefee += Gwei(max(1, state.basefee * (len(block.core.data) - SHARD_BLOCK_SIZE_TARGET) // QUOTIENT))
-    elif len(block.core.data) < SHARD_BLOCK_SIZE_TARGET:
-        state.basefee -= Gwei(max(1, state.basefee * (len(block.core.data) - SHARD_BLOCK_SIZE_TARGET) // QUOTIENT))
-    state.basefee = Gwei(max(
-        1,
-        min(
-            EFFECTIVE_BALANCE_INCREMENT // EPOCHS_PER_SHARD_PERIOD // SHARD_SLOTS_PER_BEACON_SLOT * SLOTS_PER_EPOCH,
-            state.basefee,
-        )
-    ))
-
-    # Check total bytes
-    state.total_bytes += len(block.core.data)
-    assert block.core.total_bytes == state.total_bytes
-
-    # Update in-state block header
-    state.most_recent_block_core = ShardBlockCore(
-        slot=block.core.slot,
-        beacon_chain_root=block.core.beacon_chain_root,
-        parent_root=block.core.parent_root,
-        data_root=hash_tree_root(block.core.data),
-        state_root=Hash(),
-        total_bytes=block.core.total_bytes,
-        attester_bitfield=block.core.attester_bitfield,
+```python
+def process_shard_block_header(beacon_state: BeaconState, shard_state: ShardState, block: ShardBlock) -> None:
+    # Verify the shard number
+    assert block.shard == shard_state.shard
+    # Verify the slot number
+    assert block.slot == shard_state.slot
+    # Verify the beacon chain root
+    parent_epoch = compute_epoch_of_shard_slot(shard_state.latest_block_header.slot)
```

Comment:
What is the rationale for verifying the slot from the latest block header? After a significant # of skip blocks you get an old picture. It also feels like this would be out of sync with the rest coming from the beacon state at the epoch boundaries - and then this by default making the fork choice also lag behind? Maybe I misunderstand the fork choice rule... I took it to mean you start from the latest winning/accepted crosslink root which has been calculated from the boundary of the epoch represented in the beacon block root.

Reply:
I suppose this does avoid problems right at the boundary though... which is why I assume this was there i the first place (IE at the first slot this will make you go back one epoch).

Reply:
Although this wouldn't always protect you on the 2nd shard slot of the epoch (assuming 2 shard slots per beacon slot)

Reply:
This does look wrong, thanks! 👍